oeps/oep-38-Data-Modeling.rst

```diff
@@ -0,0 +1,438 @@
+
++---------------+------------------------------------------------------------+
+
+| Status | Provisional |
+
++---------------+------------------------------------------------------------+
+
+| Type | Architecture |
+
++---------------+------------------------------------------------------------+
+
+| Created | 2018-04-19 |
+
++---------------+------------------------------------------------------------+
+
+| Resolution | - 'Original pull request' |
+
+| | - 'Update pull request' |
+
++---------------+------------------------------------------------------------+
+
+| References | - 'OEP-002' |
+
++---------------+------------------------------------------------------------+
+
+
+.. \_Original pull request:
+
+.. \_Update pull request:
+
+.. \_OEP-002:
+https://open-edx-proposals.readthedocs.io/en/latest/oep-0002-bp-repo-metadata.html
+
+Abstract
+
+========
+
+In order to obtain the highest possible value from data collected in the
+edX ecosystem, this document attempts to provide guiding thoughts and
+principles on data modeling.
+
+These data modeling guidelines are broken down into three sections:
+
+-  Guiding Principles
+
+-  Data Modeling Standards
+
+-  Best Practices
+
+Guiding Principles - This section is intended to give the reader a
+framework to think about data modeling at edX. While no specific
+instructions are given in this section, it should act as a framework to
+help the reader think about data modeling on the whole at edX.
+
+Data modeling standards - The core guidelines that everyone must adhere
+to when creating new data models at edX. These standards are designed to
+ensure edX can gain the highest value and insights from the data. They
+should be applied as the basic standard for all applications in the
+OpenedX ecosystem.
+
+Data Best Practices - These practices are designed to help teams create
+rich and efficient data models within the edX ecosystem. They are not
+standards but guidelines to help teams think about how they are storing
+data.
+
+This OEP explicitly does not cover:
+
+-  Data warehousing
+
+-  Data pipelines
+
+-  Event data
+
+Motivation
+
+==========
+
+The goal of this OEP is to provide a framework about how to think about
+data modeling at edX and provide guidelines on what should and should
+not be done when creating data models.
+
+Definitions
+
+===========
+
+BI - Business intelligence. Technologies, applications and practices for
+the collection, integration, analysis, and presentation of business
+information. The purpose of Business Intelligence is to support better
+business decision making.
+
+Composite Key - in the context of relational databases, a composite key
+is a combination of two or more columns in a table that can be used to
+uniquely identify each row in the table. Uniqueness is only guaranteed
+when the columns are combined; when taken individually, the columns do
+not guarantee uniqueness.
+
+CUD - Create, Update, Delete. These are some of the actions that may
```

Comment:
Is this intentionally omitting the Read in CRUD?

Reply:
CUD only applies to CoW databases...

Reply:
It is, There is a section below that deals with when to write history. Changes in a tables state only happens during Create, update and delete actions. Hence the omission of Read.